### DISCUSSION TITLE
What is best practice to keep both docker compose configuration up-to-date and custom configuration?

### DISCUSSION THREAD

**Discussion by archichen** [langgenius/dify#4032] [General]
I need connect dify to my notion workspace, so I change the environment in docker compose file, but I wanna keep dify up to date as well, when I use like 'git pull' to sync code with main repo, everything was overwritten, so what's best practice can do that?

**Comment by TinsFox** [comment]
My approach is to switch to a new local branch, such as local-dev, modify the changes I need in this branch, and then use docker-compose under this branch to deploy. When dify is updated, execute git rebase upstream/main under the local-dev branch.

In this way, the updates of dify can be pulled into local-dev. If dify does not make changes to docker compose, there will generally be no conflict.

My complete steps are as follows:
1. fork
2. `git clone https://github.com/<your-username>/dify`
3. `git remote add upstream https://github.com/langgenius/dify.git`
4. `git checkout -b local-dev`
5. Modify the file according to your needs
6. Submit changed files, `git add . && git commit -m "what have you done"`
7. If you know that dify has been updated, `git fetch --all && git rebase upstream/main`


hope this helps you

**Comment by hjlarry** [comment]
I think the simplest way is to use the  `git stash` method

1. git stash
2. git pull
3. git stash pop

I use this flow to upgrade from 0.6.6 to 0.6.7, it`s works fine.

**Comment by shelltea** [comment]
You can use this script：

```bash
#!/bin/bash

# 设置仓库的路径。请根据实际情况修改这个路径。
REPO_PATH="./dify"

# 进入仓库目录
cd "$REPO_PATH" || { echo "Repository path '$REPO_PATH' not found."; exit 1; }

# 检查是否存在未提交的更改
if [[ $(git status --porcelain) ]]; then
    # 暂存当前工作目录中的所有更改
    git stash push -u -m "Local changes before update $(date +%Y-%m-%d:%H:%M:%S)"
    STASH_APPLIED=1
else
    echo "No local changes to stash."
fi

# 拉取最新的代码，并尝试自动合并
git pull --rebase

# 如果之前有暂存过更改，则尝试重新应用它们
if [ "$STASH_APPLIED" == "1" ]; then
    # 应用之前暂存的更改
    if git stash pop; then
        echo "Successfully reapplied your local changes."
    else
        echo "There were conflicts when reapplying your local changes."
        exit 1
    fi
fi

echo "Repository has been updated."
```

After modifying `REPO_PATH`, save the above content to `update_dify.sh`. To use it, execute: `./update_dify.sh`
